src/shims/unix/linux/epoll.rs

```diff
@@ -95,42 +148,92 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
 
-            let events = this.project_field(&event, 0)?;
-            let events = this.read_scalar(&events)?.to_u32()?;
-            let data = this.project_field(&event, 1)?;
-            let data = this.read_scalar(&data)?;
-            let event = EpollEvent { events, data };
+            // We only support edge-triggered notification for now.
+            if events & epollet != epollet {
+                throw_unsup_format!("epoll_ctl: epollet flag must be included.");
+            }
 
-            let Some(mut epfd) = this.machine.fds.get_mut(epfd) else {
-                return Ok(Scalar::from_i32(this.fd_not_found()?));
+            let epoll_key = (weak_file_descriptor, fd);
+
+            // Check if the fd is already in the interest list.
+            if op == epoll_ctl_add {
+                if interest_list.contains_key(&epoll_key) {
+                    let eexist = this.eval_libc("EEXIST");
+                    drop(epfd);
+                    this.set_last_error(eexist)?;
+                    return Ok(Scalar::from_i32(-1));
+                }
+            } else {
+                if !interest_list.contains_key(&epoll_key) {
+                    let enoent = this.eval_libc("ENOENT");
+                    drop(epfd);
+                    this.set_last_error(enoent)?;
+                    return Ok(Scalar::from_i32(-1));
+                }
+            }
+
+            let weak_file_descriptor = file_descriptor.get_weak_file_descriptor();
+            let binding = weak_file_descriptor.get_file_description().unwrap();
+            let target_file_description = binding.borrow_mut();
+            let event = EpollEvent {
+                file_descriptor: fd,
+                weak_file_descriptor,
+                events,
+                data,
+                ready_list: Rc::clone(ready_list),
             };
-            let epfd = epfd
-                .downcast_mut::<Epoll>()
-                .ok_or_else(|| err_unsup_format!("non-epoll FD passed to `epoll_ctl`"))?;
+            // Modify the interest list entry if it already exists (epoll_ctl),
+            // and insert if it doesn't (epoll_add).
+            interest_list.entry(epoll_key).and_modify(|e| *e = event.clone()).or_insert(event);
+            // Check and update the fd ready list if there is any event.
+            drop(epfd);
+            target_file_description.check_and_update_readiness(this)?;
 
-            epfd.file_descriptors.insert(fd, event);
             Ok(Scalar::from_i32(0))
         } else if op == epoll_ctl_del {
-            let Some(mut epfd) = this.machine.fds.get_mut(epfd) else {
-                return Ok(Scalar::from_i32(this.fd_not_found()?));
-            };
-            let epfd = epfd
-                .downcast_mut::<Epoll>()
-                .ok_or_else(|| err_unsup_format!("non-epoll FD passed to `epoll_ctl`"))?;
+            let epoll_key = (weak_file_descriptor, fd);
+            if !interest_list.contains_key(&epoll_key) {
```

Comment:
```suggestion
            if interest_list.remove(&epoll_key).is_none() {
```

Removing and checking if something was removed can be done in one operation I think (didn't check the docs, so may be wrong)

Reply:
Yup, this is achievable.